packages/flutter_tools/lib/src/run_hot.dart

```diff
@@ -988,3 +922,93 @@ class HotRunner extends ResidentRunner {
     await stopEchoingDeviceLog();
   }
 }
+
+class ProjectFileInvalidator {
```

Comment:
Oh, one more question. How is/should this be tested?

Reply:
Right now it is tested via the `integration/hot_reload_test.dart`, the devicelab, and via some google3 tooling tests. I could make this injectable and add some mock filesystem tests

Reply:
Actually thanks to the memory filesystem I didn't need to be too fancy here, unit tests added as well.